File: backend/app/schemas/mobile.py
"""Pydantic schemas for Mobile PWA & Offline CRM (Phase 15)."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class OfflineQueueItemCreate(BaseModel):
    client_id: str = Field(..., max_length=64)
    resource: str = Field(..., max_length=50)
    action: str = Field(..., max_length=20)
    entity_id: str | None = None
    payload: dict[str, Any] = Field(default_factory=dict)


class OfflineQueueItemResponse(BaseModel):
    id: uuid.UUID
    client_id: str
    resource: str
    action: str
    entity_id: str | None
    payload: dict[str, Any]
    status: str
    retry_count: int
    error_message: str | None
    created_at: datetime
    processed_at: datetime | None

    model_config = {"from_attributes": True}


class OfflineQueueBatchRequest(BaseModel):
    items: list[OfflineQueueItemCreate]


class OfflineQueueBatchResponse(BaseModel):
    accepted: int
    items: list[OfflineQueueItemResponse]


class SyncResourceRequest(BaseModel):
    resources: list[str] = Field(default_factory=list)
    direction: str = "bidirectional"
    since: datetime | None = None


class SyncConflictResponse(BaseModel):
    id: uuid.UUID
    resource: str
    entity_id: str
    client_version: dict[str, Any]
    server_version: dict[str, Any]
    status: str
    resolution: str | None
    created_at: datetime
    resolved_at: datetime | None

    model_config = {"from_attributes": True}


class SyncSessionResponse(BaseModel):
    id: uuid.UUID
    direction: str
    status: str
    resources: list[str]
    items_uploaded: int
    items_downloaded: int
    conflicts_found: int
    started_at: datetime | None
    completed_at: datetime | None
    error_message: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class SyncDataResponse(BaseModel):
    session: SyncSessionResponse
    data: dict[str, list[dict[str, Any]]]
    conflicts: list[SyncConflictResponse]
    server_time: datetime


class ConflictResolveRequest(BaseModel):
    resolution: str = Field(..., pattern="^(client_wins|server_wins|merged|manual)$")
    merged_data: dict[str, Any] | None = None


class PushSubscriptionCreate(BaseModel):
    endpoint: str
    keys: dict[str, str]
    user_agent: str | None = None
    preferences: dict[str, Any] = Field(default_factory=dict)


class PushSubscriptionResponse(BaseModel):
    id: uuid.UUID
    endpoint: str
    status: str
    preferences: dict[str, Any]
    created_at: datetime
    last_used_at: datetime | None

    model_config = {"from_attributes": True}


class MobileSettingsResponse(BaseModel):
    offline_enabled: bool
    background_sync: bool
    push_enabled: bool
    auto_download: bool
    cache_resources: list[str]
    last_sync_at: datetime | None
    storage_used_bytes: int
    preferences: dict[str, Any]

    model_config = {"from_attributes": True}


class MobileSettingsUpdate(BaseModel):
    offline_enabled: bool | None = None
    background_sync: bool | None = None
    push_enabled: bool | None = None
    auto_download: bool | None = None
    cache_resources: list[str] | None = None
    storage_used_bytes: int | None = None
    preferences: dict[str, Any] | None = None


class MobileDashboardResponse(BaseModel):
    is_online: bool = True
    offline_queue_pending: int
    open_conflicts: int
    last_sync_at: datetime | None
    storage_used_bytes: int
    push_subscriptions: int
    recent_sessions: list[SyncSessionResponse]
    cacheable_resources: list[str]

File: frontend/types/mobile.ts
export type CacheResource =
  | "dashboard"
  | "companies"
  | "contacts"
  | "leads"
  | "deals"
  | "tasks"
  | "calendar"
  | "meetings"
  | "notes"
  | "documents"
  | "notifications"
  | "activities"
  | "ai_history"
  | "settings";

export interface OfflineQueueItem {
  client_id: string;
  resource: string;
  action: string;
  entity_id?: string | null;
  payload: Record<string, unknown>;
}

export interface OfflineQueueItemResponse extends OfflineQueueItem {
  id: string;
  status: string;
  retry_count: number;
  error_message: string | null;
  created_at: string;
  processed_at: string | null;
}

export interface SyncSession {
  id: string;
  direction: string;
  status: string;
  resources: string[];
  items_uploaded: number;
  items_downloaded: number;
  conflicts_found: number;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
  created_at: string;
}

export interface SyncConflict {
  id: string;
  resource: string;
  entity_id: string;
  client_version: Record<string, unknown>;
  server_version: Record<string, unknown>;
  status: string;
  resolution: string | null;
  created_at: string;
  resolved_at: string | null;
}

export interface MobileSettings {
  offline_enabled: boolean;
  background_sync: boolean;
  push_enabled: boolean;
  auto_download: boolean;
  cache_resources: string[];
  last_sync_at: string | null;
  storage_used_bytes: number;
  preferences: Record<string, unknown>;
}

export interface MobileDashboard {
  is_online: boolean;
  offline_queue_pending: number;
  open_conflicts: number;
  last_sync_at: string | null;
  storage_used_bytes: number;
  push_subscriptions: number;
  recent_sessions: SyncSession[];
  cacheable_resources: string[];
}

export interface SyncDataResponse {
  session: SyncSession;
  data: Record<string, Record<string, unknown>[]>;
  conflicts: SyncConflict[];
  server_time: string;
}

export interface PushSubscriptionInfo {
  id: string;
  endpoint: string;
  status: string;
  preferences: Record<string, unknown>;
  created_at: string;
  last_used_at: string | null;
}

export interface CachedEntity {
  id: string;
  resource: CacheResource;
  data: Record<string, unknown>;
  updated_at: string;
  synced_at: string;
}

export interface StorageStats {
  usedBytes: number;
  itemCount: number;
  resources: Record<string, number>;
}
